### DISCUSSION TITLE
What should I do on writing plugins when each document's metadata affects other documents?

### DISCUSSION THREAD

**Discussion by McDic** [mkdocs/mkdocs#3553] [Q&A] [answered]
I was writing a plugin for my own blog. I was trying to provide list of recently edited posts on index page and I decided to generate special HTML for index page only. But since I am not developing a theme(I use [mkdocs-material](https://squidfunk.github.io/mkdocs-material/)), I decided to alter markdown text for easier development.

According to [Mkdocs plugin development documentation](https://www.mkdocs.org/dev-guide/plugins), html page is constructed after `on_page_markdown` and modified by `on_page_content`, **independently on each page**.

![스크린샷 2024-01-18 224223](https://github.com/mkdocs/mkdocs/assets/19789019/c32e619f-2a30-43e9-b64a-f8c5cf845531)

It seems there isn't any straightforward way to make kind of synchronized global event during some step of `populate_page` process. Is `mkdocs` implemented in this way because of page building performance, or is there any other reason?

I think letting post metadata affect global navigation and optionally some other pages is a valid usecase for many users. So my final questions are..

- What should I do as easy bandaid patch for this?
- Will `mkdocs` provide easier interface to make custom global events and synchronizations at some point?

**Comment by oprypin** [comment, marked as answer]
You can make a particular page be rendered last. That way you can ensure all other pages' events have run by the time you reach that one.

Inside an `on_files` handler do this:

```python
file = files.get_file_from_path('index.md')
files.remove(file)
files.append(file)
```

Or just append a fresh file without removing anything in the first place

**Reply by oprypin** [reply]
A plugin that does this: https://github.com/squidfunk/mkdocs-material/blob/44982451f7759ed65e31d1151b3c49566b8c6ca0/material/plugins/tags/plugin.py#L129

**Reply by McDic** [reply]
Thank you. This can be a quick patch, although this kind of way only works when file dependencies form a DAG. Not sure if that manual manipulation of `files` is intended and encouraged by `mkdocs`. If that is the case, I guess this is better being documented.

**Comment by oprypin** [comment]
To make a page's content affect navigation, that needs to be done in `on_files`. Only the raw content of files is available though.

**Reply by oprypin** [reply]
A plugin that does this: https://github.com/oprypin/mkdocs-literate-nav/blob/fb047ef72f1a65ca305852933c2a7f100a7279d7/mkdocs_literate_nav/plugin.py#L36
